include/base/ghosting_functor.h

```diff
@@ -159,11 +160,29 @@ class GhostingFunctor : public ReferenceCountedObject<GhostingFunctor>
    */
   GhostingFunctor() {}
 
+  /**
+   * Copy Constructor
+   */
+  GhostingFunctor(const GhostingFunctor & other) : ReferenceCountedObject<GhostingFunctor>(other) {}
+
   /**
    * Virtual destructor; this is an abstract base class.
    */
   virtual ~GhostingFunctor() {}
 
+  /**
+   * A clone() is needed because GhostingFunctor can not be shared between
+   * different meshes. The operations in  GhostingFunctor are mesh dependent.
+   */
+  virtual std::unique_ptr<GhostingFunctor> clone () const
+  { libmesh_not_implemented(); }
```

Comment:
This is your problem.  As soon as we hit a subclass that hasn't be modified in sync with this API change and the next, we'll scream and die.  Better to make clone() a pure virtual function so at least the failure happens at compile time (in this case, as soon as you hit include/base/sibling_coupling.h) instead of runtime!

On the other hand, it occurs to me that the MOOSE update here is going to be a lot smoother than I'd expected.  There's nothing stopping us from implementing clone() in the RelationshipManager subclasses there *before* we do the libMesh update, and the result will be compatible with libMesh both pre- and post- this PR; we just can't mark those implementations as override until after this PR gets merged.

Reply:
Oh, and although of course the MOOSE-based CI in this PR won't pass until those subclass clone methods are merged to MOOSE, let's save that until after we've got the code here passing libMesh-only tests and reviewed; it would be silly to merge something into MOOSE only to risk realizing afterwards that we need a slight design change, like my suggestion about set_mesh() below.

Oh, and speak of the devil:  We should have a set_mesh() marked virtual in this class, shouldn't we, even though we'll be able to implement it here and that implementation will be good enough for most/all subclasses today?  There may be subclasses that want to override that implementation to re-cache mesh-based data when needed, and it's not a function that will ever be called in any inner loops so the virtual overhead should be negligible.

Reply:
OK, we have to implement clone correctly in libmesh before my MOOSE PR can get in since a mesh could have both RMs and libmesh style ghosting functors. I can not miss any of them in the displaced mesh.

Reply:
Oh, never mind.

Reply:
In order to have a back-compatibility, I think we might do this in `ghosting_functor.h `:

```
  virtual std::unique_ptr<GhostingFunctor> clone () const
  {  return nullprt; }
```

and then do this in `meh_base.C`:


```
   for (const auto & gf : other_mesh._ghosting_functors )
    {
      std::shared_ptr<GhostingFunctor> clone_gf = gf->clone();
      if (clone_gf)
        {
           clone_gf->set_mesh(this);
           add_ghosting_functor(clone_gf);
        } else 
       {
           add_ghosting_functor(gf);
       }
    }
```

Reply:
Once the integration is done, we could come back and change this function to pure virtual

Reply:
I think the cleverness here means we can afford to wait on going pure virtual - let's let a release get out where the old API just screams ```libmesh_deprecated();``` then we can go to pure virtual for the release after that.